### PR TITLE
[merged] Atomic/push.py: Implement push with signing

### PR DESCRIPTION
**Pull request by baude**
Enabling signing and pushing at the same time.  At the time of
this writing, we cannot verify that the atomic targets are
working but the code is there.  THis is enabled with the
--type atomic switch. Otherwise, signatures are written locally
and the image is pushed.

### CODE REVIEW

File: Atomic/push.py [hunk 1]
```diff
@@ -133,19 +142,41 @@ def push(self):
...
+
+            # We must push the file to the registry first prior to performing a
+            # local signature because the manifest file must be on the registry
+            return_code = util.skopeo_copy(local_image, remote_image, debug=self.args.debug,
+                                           sign_by=self.args.sign_by if not sign_local else None, insecure=insecure)
+
+            if return_code != 0:
+                raise ValueError("Pushing {} failed.".format(self.image))
+
+            util.write_out("Pushed: {}".format(self.image))
```
Comment (rhatdan) [inline Atomic/push.py]: No debug statements.

File: Atomic/push.py [hunk 2]
```diff
@@ -133,19 +142,41 @@ def push(self):
...
+            local_image = "docker-daemon:{}".format(self.image)
+            if self.args.reg_type == "atomic":
+                remote_image = "atomic:{}".format(self.image)
+                sign_local = False
+            else:
+                remote_image = "docker://{}".format(self.image)
+
+            sign = True if self.args.sign_by else False
+            if sign:
+                util.write_out("\nSigning with '{}'\n".format(self.args.sign_by))
```
Comment (rhatdan) [inline Atomic/push.py]: No debug statements.

File: Atomic/push.py [hunk 3]
```diff
@@ -133,19 +142,41 @@ def push(self):
...
-                elif status.startswith("Pushing"):
-                    util.write_out("Pushing: " + bar['id'])
-
-                prevstatus = status
+            reg, _, tag = util.decompose(self.image)
+            if not tag:
+                raise ValueError("The image being pushed must have a tag")
+
+            if self.args.password and self.args.username:
+                print(self.d.login(username=self.args.username, password=self.args.password, registry=reg))
```
Comment (rhatdan) [inline Atomic/push.py]: No debug messages.

File: Atomic/sign.py [hunk 4]
```diff
@@ -25,24 +30,38 @@ def cli(subparser):
                        help=_("Define an alternate directory to store signatures"))
 
 class Sign(Atomic):
-    def sign(self):
+    def __init__(self):
+        super(Sign, self).__init__()
+
+    def sign(self, debug=None, signer=None, in_signature_path=None, images=None):
```
Comment (rhatdan) [inline Atomic/sign.py]: Make debug a param of __init__, Maybe others

File: Atomic/push.py [hunk 5]
```diff
@@ -133,19 +142,41 @@ def push(self):
...
+            if return_code != 0:
+                raise ValueError("Pushing {} failed.".format(self.image))
+
+            util.write_out("Pushed: {}".format(self.image))
+
+            if sign_local and sign:
+                if self.args.debug:
+                    util.write_out("Need to create a local signature")
+                atomic_sign = Sign()
+                atomic_sign.sign(debug=self.args.debug, signer=self.args.sign_by,
```
Comment (rhatdan) [inline Atomic/push.py]: We could just do
atomic_sign.args.sign_by=self.args.sign_by
...

Rather then adding all of these parameters.

**Comment by rhatdan** [comment]
Also please fix the test failures.

File: Atomic/push.py (hunk 5 above)
Comment (baude) [inline Atomic/push.py]: I dont think that will work because the self.args differ between the commands.

File: Atomic/push.py (hunk 3 above)
Comment (baude) [inline Atomic/push.py]: Fixed.

File: Atomic/push.py (hunk 1 above)
Comment (baude) [inline Atomic/push.py]: Wrapped with if debug

File: Atomic/push.py (hunk 2 above)
Comment (baude) [inline Atomic/push.py]: wrapped with if debug

File: Atomic/push.py [hunk 6]
```diff
@@ -71,12 +75,17 @@ def cli(subparser):
...
     def push(self):
         self.ping()
-        prevstatus = ""
+        if self.args.debug:
+            util.write_out(str(self.args))
+
+        # This allows a user to turn off signing when a global
+        # sign_by has been defined in /etc/atomic.conf and saves
+        # us from having to define something like --no-sign
+        if self.args.sign_by == "None":
```
Comment (rhatdan) [inline Atomic/push.py]: Who is passing in "None" string?

File: Atomic/pull.py [hunk 7]
```diff
@@ -18,20 +19,27 @@ def cli(subparser):
...
+    pullp.add_argument("-t", "--type", dest="reg_type", default=None,
+                       help=_("Push to an alternative registry type."))
     pullp.add_argument("image", help=_("image id"))
 
 
 class Pull(Atomic):
     def pull_docker_image(self):
-        _, _, tag = decompose(self.args.image)
+        registry, _, tag = decompose(self.args.image)
+        insecure = True if is_insecure_registry(self.d.info()['RegistryConfig'], remove_port(registry)) else False
```
Comment (rhatdan) [inline Atomic/pull.py]: The remove_port() here seems like a strange function name to call?

File: Atomic/pull.py [hunk 8]
```diff
@@ -18,20 +19,27 @@ def cli(subparser):
...
         # If no tag is given, we assume "latest"
         tag = tag if tag != "" else "latest"
-        fq_name = skopeo_inspect("docker://{}".format(self.args.image))['Name']
+        if self.args.reg_type == "atomic":
+            pull_uri = 'atomic:'
```
Comment (rhatdan) [inline Atomic/pull.py]: The atomic does not use the atomic:// path?

File: Atomic/push.py (hunk 5 above)
Comment (rhatdan) [inline Atomic/push.py]: You just have to make sure that Sign has them all defined in its init.

File: Atomic/util.py [hunk 9]
```diff
@@ -635,3 +647,64 @@ def get_signature_read_path(reg_info):
     # Return the defined path for where signatures should be read
     # or none if no entry is found
     return reg_info.get('sigstore', None)
+
+def remove_port(_input):
```
Comment (rhatdan) [inline Atomic/util.py]: Maybe this should be called strip_port_text or something like that, I see remove_port, and I think you are changing something in the configuration.

File: Atomic/push.py (hunk 6 above)
Comment (baude) [inline Atomic/push.py]: you can pass None to turn off signing if a default signer is defined in /etc/atomic.conf like the comment above says

File: Atomic/pull.py (hunk 7 above)
Comment (baude) [inline Atomic/pull.py]: ok def strip_port() now.

File: Atomic/pull.py (hunk 8 above)
Comment (baude) [inline Atomic/pull.py]: nope, it is defined differently by skopeo than the others.

File: Atomic/util.py (hunk 9 above)
Comment (baude) [inline Atomic/util.py]: yeah, took your suggestion.

File: Atomic/pull.py (hunk 8 above)
Comment (rhatdan) [inline Atomic/pull.py]: @runcom can you fix this?  Should follow the URI format.

File: Atomic/push.py (hunk 6 above)
Comment (rhatdan) [inline Atomic/push.py]: Ok, now I get it.

File: Atomic/sign.py (hunk 4 above)
Comment (baude) [inline Atomic/sign.py]: @rhatdan I will push a new update here shortly.  I have removed all the named args that overlap with self.args between push and sign.  That only leaves images and in_signature_path which are not in push.

File: Atomic/pull.py (hunk 8 above)
Comment (runcom) [inline Atomic/pull.py]: @rhatdan we stripped the double slash long time ago when designing system containers IIRC, any reason we want it again? @mtrmac do you remember by any chance?

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

File: Atomic/pull.py (hunk 8 above)
Comment (rhatdan) [inline Atomic/pull.py]: Just seems weird not to have it.

Comment (mtrmac) [inline Atomic/pull.py]: The original motivation was simply that the `//` is too much mandatory=meaningless noise and extra typing.

Strictly speaking URI format is schema:schema-specific something; `//` is used only when the part starts with an “authority” (server name/address), and our original design didn’t have any hostnames in `atomic:`.  Now that `docker:` and `atomic:` use the same docker API endpoint + repo etc. semantics, being consistent about `//` would make more sense…

… but then I’ve been tempted for a long time to make the `//` in `docker://` optional :)

Comment (mtrmac) [inline Atomic/pull.py]: (… actually, the shorthand notation like `docker://busybox` is violating the rules.)

Comment (rhatdan) [inline Atomic/pull.py]: Does this mean it will break?  I don't care as long as they are consistent.

Comment (mtrmac) [inline Atomic/pull.py]: We would just accept but not require the `//`. No existing users should break.

**Comment by cgwalters** [comment]
Note this required http://pkgs.fedoraproject.org/cgit/rpms/atomic.git/commit/?id=099aba986bf53edb082871db7c9519f67710b4c5

**Pull request closed without merging**